python/ray/data/_internal/execution/autoscaler/default_autoscaler.py

```diff
@@ -52,74 +59,83 @@ def _calculate_actor_pool_util(self, actor_pool: AutoscalingActorPool):
-        if op_state.total_enqueued_input_bundles() <= actor_pool.num_free_task_slots():
-            return False
-        # Determine whether to scale up based on the actor pool utilization.
-        util = self._calculate_actor_pool_util(actor_pool)
-        return util > self._actor_pool_scaling_up_threshold
+            return _AutoscalingAction.SCALE_DOWN, "pool exceeding max size"
 
-    def _actor_pool_should_scale_down(
-        self,
-        actor_pool: AutoscalingActorPool,
-        op: "PhysicalOperator",
-        op_state: "OpState",
-    ):
-        # Scale down, if the op is completed or no more inputs are coming.
-        if op.completed() or (op_state.total_enqueued_input_bundles() == 0):
-            return True
-        if actor_pool.current_size() > actor_pool.max_size():
-            # Scale down, if the actor pool is above max size.
-            return True
-        elif actor_pool.current_size() <= actor_pool.min_size():
-            # Do not scale down, if the actor pool is already at min size.
-            return False
-        # Determine whether to scale down based on the actor pool utilization.
+        # Determine whether to scale up based on the actor pool utilization.
         util = self._calculate_actor_pool_util(actor_pool)
-        return util < self._actor_pool_scaling_down_threshold
+        if util >= self._actor_pool_scaling_up_threshold:
+            # Do not scale up if either
+            #   - Previous scale up has not finished yet
+            #   - Actor Pool is at max size already
+            #   - Op is throttled (ie exceeding allocated resource quota)
+            #   - Actor Pool has sufficient amount of slots available to handle
+            #   pending tasks
+            if actor_pool.num_pending_actors() > 0:
+                return _AutoscalingAction.NO_OP, "pending actors"
+            elif actor_pool.current_size() >= actor_pool.max_size():
+                return _AutoscalingAction.NO_OP, "reached max size"
+            if not op_state._scheduling_status.under_resource_limits:
+                return _AutoscalingAction.NO_OP, "operator exceeding resource quota"
+            elif (
+                op_state.total_enqueued_input_bundles()
+                <= actor_pool.num_free_task_slots()
+            ):
+                return _AutoscalingAction.NO_OP, (
+                    f"pool has sufficient task slots remaining: "
+                    f"enqueued inputs {op_state.total_enqueued_input_bundles()} <= "
+                    f"free slots {actor_pool.num_free_task_slots()})"
+                )
+
+            return (
+                _AutoscalingAction.SCALE_UP,
+                f"utilization of {util} >= {self._actor_pool_scaling_up_threshold}",
+            )
+        elif util <= self._actor_pool_scaling_down_threshold:
+            if not actor_pool.can_scale_down():
+                return _AutoscalingAction.NO_OP, "debounced"
```

Comment:
nit, the method name `can_scale_down` doesn't suggest it's because of debouncing

Reply:
Updated